polaris-core/src/main/java/org/apache/polaris/core/persistence/cache/EntityCacheGrantManager.java

```diff
@@ -0,0 +1,247 @@
+                    List::add,
+                    (l, r) -> {
+                      l.addAll(r);
+                      return l;
+                    },
+                    Collector.Characteristics.IDENTITY_FINISH));
+
+    // If the securable is the root container, then we need to add a grant record for the
+    // service_admin PrincipalRole, which is the only role that has the SERVICE_MANAGE_ACCESS
+    // privilege on the root
+    if (lookupResult
+        .getCacheEntry()
+        .getEntity()
+        .getName()
+        .equals(PolarisEntityConstants.getRootContainerName())) {
+      if (serviceAdminEntity == null || serviceAdminRootContainerGrant == null) {
+        EntityCacheLookupResult serviceAdminRole =
+            entityCache.getOrLoadEntityByName(
+                callCtx,
+                new EntityCacheByNameKey(
+                    0L,
+                    0L,
+                    PolarisEntityType.PRINCIPAL_ROLE,
+                    PolarisEntityConstants.getNameOfPrincipalServiceAdminRole()));
+        if (serviceAdminRole == null || serviceAdminRole.getCacheEntry() == null) {
+          LOGGER.error("Failed to resolve service_admin PrincipalRole");
+          return new LoadGrantsResult(BaseResult.ReturnStatus.GRANT_NOT_FOUND, null);
+        }
+        serviceAdminRootContainerGrant =
+            new PolarisGrantRecord(
+                0L,
+                0L,
+                serviceAdminRole.getCacheEntry().getEntity().getCatalogId(),
+                serviceAdminRole.getCacheEntry().getEntity().getId(),
+                PolarisPrivilege.SERVICE_MANAGE_ACCESS.getCode());
+        serviceAdminEntity = serviceAdminRole.getCacheEntry().getEntity();
+      }
+      grantRecords.add(serviceAdminRootContainerGrant);
+      granteeList.add(serviceAdminEntity);
+    }
+    return new LoadGrantsResult(
+        lookupResult.getCacheEntry().getEntity().getGrantRecordsVersion(),
+        grantRecords,
+        granteeList);
+  }
+
+  @Override
+  public @NotNull LoadGrantsResult loadGrantsToGrantee(
+      PolarisCallContext callCtx, long granteeCatalogId, long granteeId) {
+    EntityCacheLookupResult lookupResult =
+        entityCache.getOrLoadEntityById(callCtx, granteeCatalogId, granteeId);
+    if (lookupResult == null || lookupResult.getCacheEntry() == null) {
+      return new LoadGrantsResult(BaseResult.ReturnStatus.GRANT_NOT_FOUND, null);
+    }
+    List<PolarisBaseEntity> granteeList =
+        lookupResult.getCacheEntry().getGrantRecordsAsGrantee().stream()
+            .map(
+                gr ->
+                    entityCache.getOrLoadEntityById(
+                        callCtx, gr.getSecurableCatalogId(), gr.getSecurableId()))
+            .takeWhile(
+                lr -> {
+                  if (lr == null || lr.getCacheEntry() == null) {
+                    throw new IllegalStateException("Grantee not found");
+                  }
+                  return true;
+                })
```

Comment:
we're not actually creating a logical "sub-stream" here, but fail on a certain condition... IMHO `.filter()` would be more intuitive.

Reply:
Converted to a filter and changed to throw exception if the size doesn't match the expected